client/config_client.go

```diff
@@ -0,0 +1,382 @@
+	Get(ctx context.Context, v *configpb.Version, component, componentID string) (*configpb.Status, *configpb.Version, string, error)
+	Update(ctx context.Context, v *configpb.Version, kind *configpb.ConfigKind, entries []*configpb.ConfigEntry) (*configpb.Status, *configpb.Version, error)
+	Delete(ctx context.Context, v *configpb.Version, kind *configpb.ConfigKind) (*configpb.Status, error)
+	// Close closes the client.
+	Close()
+}
+
+type configClient struct {
+	urls      []string
+	clusterID uint64
+	connMu    struct {
+		sync.RWMutex
+		clientConns map[string]*grpc.ClientConn
+		leader      string
+	}
+
+	checkLeaderCh chan struct{}
+
+	wg     sync.WaitGroup
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	security SecurityOption
+}
+
+// NewConfigClient creates a PD configuration client.
+func NewConfigClient(pdAddrs []string, security SecurityOption) (ConfigClient, error) {
+	return NewConfigClientWithContext(context.Background(), pdAddrs, security)
+}
+
+// NewConfigClientWithContext creates a PD configuration client with the context.
+func NewConfigClientWithContext(ctx context.Context, pdAddrs []string, security SecurityOption) (ConfigClient, error) {
+	log.Info("[pd] create pd configuration client with endpoints", zap.Strings("pd-address", pdAddrs))
+	ctx1, cancel := context.WithCancel(ctx)
+	c := &configClient{
+		urls:          addrsToUrls(pdAddrs),
+		checkLeaderCh: make(chan struct{}, 1),
+		ctx:           ctx1,
+		cancel:        cancel,
+		security:      security,
+	}
+	c.connMu.clientConns = make(map[string]*grpc.ClientConn)
+
+	if err := c.initRetry(c.initClusterID); err != nil {
+		return nil, err
+	}
+	if err := c.initRetry(c.updateLeader); err != nil {
+		return nil, err
+	}
+	log.Info("[pd] init cluster id", zap.Uint64("cluster-id", c.clusterID))
+
+	c.wg.Add(1)
+	go c.leaderLoop()
+
+	return c, nil
+}
+
+func (c *configClient) updateURLs(members []*pdpb.Member) {
+	urls := make([]string, 0, len(members))
+	for _, m := range members {
+		urls = append(urls, m.GetClientUrls()...)
+	}
+	c.urls = urls
+}
+
+func (c *configClient) initRetry(f func() error) error {
+	var err error
+	for i := 0; i < maxInitClusterRetries; i++ {
+		if err = f(); err == nil {
+			return nil
+		}
+		time.Sleep(time.Second)
+	}
+	return errors.WithStack(err)
+}
+
+func (c *configClient) initClusterID() error {
+	ctx, cancel := context.WithCancel(c.ctx)
+	defer cancel()
+	for _, u := range c.urls {
+		timeoutCtx, timeoutCancel := context.WithTimeout(ctx, pdTimeout)
+		members, err := c.getMembers(timeoutCtx, u)
+		timeoutCancel()
+		if err != nil || members.GetHeader() == nil {
+			log.Warn("[pd] failed to get cluster id", zap.String("url", u), zap.Error(err))
+			continue
+		}
+		c.clusterID = members.GetHeader().GetClusterId()
+		return nil
+	}
+	return errors.WithStack(errFailInitClusterID)
+}
+
+func (c *configClient) updateLeader() error {
```

Comment:
Maybe we add it to PD client so we don't have to manage connections in both places.

Reply:
@overvenus PTAL

Reply:
OK